### DISCUSSION TITLE
use router.push with a URL object without re-rendering page

### DISCUSSION THREAD

**Discussion by ethanryan** [vercel/next.js#23172] [Help]
We are in the process of replacing Friday's next-routes with Next.js routing because Friday's next-routes is no longer being actively maintained: https://github.com/fridays/next-routes/issues/334

We are using `router.push` with a URL object:

https://nextjs.org/docs/api-reference/next/router#with-url-object

Unfortunately, even with `{shallow: true}`, `router.push` causes the page to re-render whenever the query is updated. For example:

```
router.push(
{
  pathname: `/[region]/[activity]/search`,
  query: { ...updatedQuery, region: region, activity: activity },
}, 
 undefined, 
 { shallow: true }
)
```

A comment from a related discussion says:

> Take a look at the docs: https://nextjs.org/docs/routing/shallow-routing
> 
> By definition of the useRouter behaviour, there has to be a re-render to update the router.query object. shallow will only disable any data fetching mechanisms like getStaticProps or getInitialProps

via: https://github.com/Corcoran-Labs/corcoran-redesign-next/pull/2401/files#r595200399

This discussion, "Ability to push changes to url without re-rendering": https://github.com/vercel/next.js/discussions/18072#discussioncomment-109059 suggests using: `window.history.replaceState(window.history.state, "", href)` instead of `router.push` to update the url without causing the page to re-render. But it seems we need an href sting in order to do this, not a URL object.

The page re-render on our search page creates a bad user experience. For example, we have a dropdown with multiple input fields so a user can enter a minimum and maximum price. But when using `router.push`, a user can't type in a price minimum and price maximum without opening the dropdown twice, because typing in a number updates the url, which causes the page to re-render and closes the dropdown.

We are reverting to using Friday's next-routes `Router.pushRoute` for now.

I've been looking into Friday's next-routes code to try to replicate their pushRoute behavior: 

https://github.com/fridays/next-routes/blob/08e8bbeaa76c5ec194abea388a193bedac7f35fc/src/index.js#L172

Do we need to call encodeURIComponent to create an href string so we could use window.history.replaceState?

Why doesn't Next.js's routing allow us to update the url without it triggering a re-render of the page?

This is especially puzzling because the docs for Friday's next-routes Router.pushRoute says: 

> It generates the URLs and calls next/router

via: https://github.com/fridays/next-routes#router-example

Any suggestions for fixing this would be a huge help. Thanks!

**Comment by akellbl4** [comment]
Hi there!
The first argument is `href` that represents a dynamic route. The second argument is `as` that represents the actual route to which a user will be navigated.
As I can see you put the second argument to push as `undefined` and I can suggest doing it in a different way.

```js
router.push(
  {
    pathname: `/[region]/[activity]/search`,
    query: { ...updatedQuery, region, activity },
  }, 
  {
    pathname: `/${region}/${activity}/search`,
    query: updatedQuery,
  }, 
  { shallow: true }
)
```
